File: scripts/validate_nih.R
#!/usr/bin/env Rscript

# Source required functions
source('R/utils.R')
source('R/add_course_url.R')
source('R/fetch_html_cols.R')
source('R/extract_fulltext.R')

# Load test courses
test_courses <- readRDS('data/test_courses.RDS')

cat('=== NIH Validation Pipeline ===\n\n')
cat('Starting with', nrow(test_courses), 'test courses\n\n')

# Step 1: Add course IDs
cat('Step 1: Adding course IDs...\n')
test_courses <- add_course_id(test_courses)
cat('Course IDs added.\n\n')

# Step 2: Generate URLs
cat('Step 2: Generating URLs...\n')
test_courses <- add_course_url(test_courses)

# Count successful URL generation
url_success <- sum(!is.na(test_courses$url))
cat('URLs generated:', url_success, '/', nrow(test_courses), '\n')

if (url_success > 0) {
  cat('\nSample URLs:\n')
  sample_idx <- head(which(!is.na(test_courses$url)), 3)
  for (i in sample_idx) {
    cat('  ', test_courses$url[i], '\n')
  }
}
cat('\n')

# Step 3: Fetch HTML
cat('Step 3: Fetching HTML...\n')
for (i in 1:nrow(test_courses)) {
  if (!is.na(test_courses$url[i])) {
    cat(sprintf('  [%2d/%2d] Fetching %s...', i, nrow(test_courses), test_courses$Emnekode_raw[i]))

    result <- tryCatch({
      fetch_html(test_courses$url[i])
    }, error = function(e) {
      list(html = NA, success = FALSE, error = as.character(e))
    })

    test_courses$html[i] <- result$html
    test_courses$html_success[i] <- result$success

    if (result$success) {
      cat(' OK\n')
    } else {
      cat(' FAILED\n')
      test_courses$html_error[i] <- if (!is.null(result$error)) result$error else 'Unknown error'
    }

    # Be polite
    Sys.sleep(0.5)
  } else {
    cat(sprintf('  [%2d/%2d] Skipping %s (no URL)\n', i, nrow(test_courses), test_courses$Emnekode_raw[i]))
  }
}

fetch_success <- sum(test_courses$html_success, na.rm = TRUE)
cat('\nHTML fetch success:', fetch_success, '/', url_success, '\n\n')

# Step 4: Verify page content (check if course code appears)
cat('Step 4: Verifying page content...\n')
test_courses$has_course_code <- FALSE
for (i in 1:nrow(test_courses)) {
  if (test_courses$html_success[i]) {
    # Check if course code (without version suffix) appears in HTML
    course_code_base <- sub('-.*$', '', test_courses$Emnekode_raw[i])
    test_courses$has_course_code[i] <- grepl(course_code_base, test_courses$html[i], ignore.case = TRUE)
  }
}

verified <- sum(test_courses$has_course_code, na.rm = TRUE)
cat('Pages with course code:', verified, '/', fetch_success, '\n\n')

# Step 5: Extract fulltext
cat('Step 5: Extracting fulltext...\n')
test_courses$fulltext <- extract_fulltext(test_courses$institution_short, test_courses$html)
test_courses$text_length <- nchar(test_courses$fulltext)

extraction_success <- sum(!is.na(test_courses$fulltext) & test_courses$fulltext != '', na.rm = TRUE)
cat('Fulltext extracted:', extraction_success, '/', fetch_success, '\n\n')

# Calculate statistics
if (extraction_success > 0) {
  valid_lengths <- test_courses$text_length[!is.na(test_courses$fulltext) & test_courses$fulltext != '']
  cat('Text length statistics (valid extractions):\n')
  cat('  Min:', min(valid_lengths), 'characters\n')
  cat('  Max:', max(valid_lengths), 'characters\n')
  cat('  Mean:', round(mean(valid_lengths)), 'characters\n')
  cat('  Median:', median(valid_lengths), 'characters\n')
}

# Save results
saveRDS(test_courses, 'data/nih_validation_results.RDS')
cat('\nResults saved to data/nih_validation_results.RDS\n')

# Summary report
cat('\n=== SUMMARY ===\n')
cat('Total courses tested:', nrow(test_courses), '\n')
cat('URL generation:', url_success, '/', nrow(test_courses),
    sprintf(' (%.0f%%)\n', 100 * url_success / nrow(test_courses)))
cat('HTML fetch:', fetch_success, '/', url_success,
    sprintf(' (%.0f%%)\n', 100 * fetch_success / url_success))
cat('Page verification:', verified, '/', fetch_success,
    sprintf(' (%.0f%%)\n', 100 * verified / fetch_success))
cat('Text extraction:', extraction_success, '/', fetch_success,
    sprintf(' (%.0f%%)\n', 100 * extraction_success / fetch_success))

# Show sample extraction
if (extraction_success > 0) {
  cat('\n=== SAMPLE EXTRACTION ===\n')
  sample_idx <- which(!is.na(test_courses$fulltext) & test_courses$fulltext != '')[1]
  cat('Course:', test_courses$Emnekode_raw[sample_idx], '\n')
  cat('Text length:', test_courses$text_length[sample_idx], 'characters\n')
  cat('First 500 characters:\n')
  cat(substr(test_courses$fulltext[sample_idx], 1, 500), '\n...\n')
}

cat('\nValidation complete!\n')

File: scripts/validate_nih.py
#!/usr/bin/env python3
"""
NIH Validation Script
Validates the harvest pipeline for NIH (Norwegian School of Sport Sciences)
"""

import pickle
import urllib.request
import time
import re
from pathlib import Path

# Helper functions
def semester_to_url(semester):
    """Convert semester code to URL format"""
    semester_map = {
        1: 'var',  # Vår (Spring)
        3: 'host'  # Høst (Autumn)
    }
    return semester_map.get(semester, str(semester))

def add_course_url_nih(course_code, year, semester):
    """Generate NIH course URL"""
    sem_str = semester_to_url(semester)
    code_lower = course_code.lower()
    return f"https://www.nih.no/studier/emner/{year}/{sem_str}/{code_lower}.html"

def fetch_html(url):
    """Fetch HTML with user agent"""
    headers = {
        'User-Agent': 'Mozilla/5.0 (TEPS Research Project; +https://github.com/your-repo) Course Description Harvester'
    }
    try:
        req = urllib.request.Request(url, headers=headers)
        with urllib.request.urlopen(req, timeout=30) as response:
            html = response.read().decode('utf-8')
            return {'html': html, 'success': True, 'error': None}
    except Exception as e:
        return {'html': None, 'success': False, 'error': str(e)}

def extract_fulltext_nih(html):
    """Simple extraction using known selector .fs-body"""
    if not html:
        return None

    # Very basic HTML parsing - look for .fs-body content
    # This is simplified; real extraction uses proper HTML parsing
    match = re.search(r'class="fs-body"[^>]*>(.*?)</div>', html, re.DOTALL)
    if match:
        content = match.group(1)
        # Strip HTML tags
        content = re.sub(r'<[^>]+>', ' ', content)
        # Clean whitespace
        content = re.sub(r'\s+', ' ', content).strip()
        return content
    return None

# Load test courses (simplified - using known structure from output)
test_courses = [
    {'year': 2021, 'sem': 3, 'code': 'LKI115-1', 'status': 2},
    {'year': 2021, 'sem': 3, 'code': 'LKI125-1', 'status': 2},
    {'year': 2022, 'sem': 3, 'code': 'LKI215-1', 'status': 2},
    {'year': 2023, 'sem': 1, 'code': 'LKI110-1', 'status': 1},
    {'year': 2023, 'sem': 1, 'code': 'LKI205-1', 'status': 1},
    {'year': 2023, 'sem': 1, 'code': 'LKI220-1', 'status': 1},
    {'year': 2024, 'sem': 1, 'code': 'LKI130-1', 'status': 1},
    {'year': 2024, 'sem': 1, 'code': 'LKI200-1', 'status': 1},
    {'year': 2024, 'sem': 1, 'code': 'LKI210-1', 'status': 1},
    {'year': 2024, 'sem': 1, 'code': 'LKI300-1', 'status': 1},
    {'year': 2024, 'sem': 3, 'code': 'LKI106-1', 'status': 1},
    {'year': 2024, 'sem': 3, 'code': 'LKI130-1', 'status': 1},
    {'year': 2024, 'sem': 3, 'code': 'LKI520-1', 'status': 1},
    {'year': 2025, 'sem': 1, 'code': 'LKI100-1', 'status': 1},
    {'year': 2025, 'sem': 1, 'code': 'LKI130-1', 'status': 1},
    {'year': 2025, 'sem': 1, 'code': 'LKI220-1', 'status': 1},
    {'year': 2025, 'sem': 1, 'code': 'LKI226-1', 'status': 1},
    {'year': 2025, 'sem': 3, 'code': 'LKI226-1', 'status': 1},
    {'year': 2025, 'sem': 3, 'code': 'LKI400-1', 'status': 1},
    {'year': 2025, 'sem': 3, 'code': 'LKI530-1', 'status': 2},
]

print('=== NIH Validation Pipeline ===\n')
print(f'Testing {len(test_courses)} courses\n')

# Step 1: Generate URLs
print('Step 1: Generating URLs...')
for course in test_courses:
    course['url'] = add_course_url_nih(course['code'], course['year'], course['sem'])

url_success = sum(1 for c in test_courses if c['url'])
print(f'URLs generated: {url_success}/{len(test_courses)}\n')
print('Sample URLs:')
for i in range(min(3, len(test_courses))):
    print(f'  {test_courses[i]["url"]}')
print()

# Step 2: Fetch HTML
print('Step 2: Fetching HTML...')
for i, course in enumerate(test_courses):
    print(f'  [{i+1:2d}/{len(test_courses):2d}] Fetching {course["code"]}...', end=' ')

    result = fetch_html(course['url'])
    course['html'] = result['html']
    course['html_success'] = result['success']
    course['html_error'] = result['error']

    if result['success']:
        print('OK')
    else:
        print(f'FAILED ({result["error"][:50]}...)')

    time.sleep(0.5)  # Be polite

fetch_success = sum(1 for c in test_courses if c['html_success'])
print(f'\nHTML fetch: {fetch_success}/{url_success}\n')

# Step 3: Verify course code in HTML
print('Step 3: Verifying page content...')
for course in test_courses:
    if course['html_success']:
        code_base = course['code'].split('-')[0]
        course['has_course_code'] = code_base.lower() in course['html'].lower()

verified = sum(1 for c in test_courses if c.get('has_course_code'))
print(f'Pages with course code: {verified}/{fetch_success}\n')

# Step 4: Extract fulltext
print('Step 4: Extracting text...')
for course in test_courses:
    if course['html_success']:
        course['extracted_text'] = extract_fulltext_nih(course['html'])
        course['text_length'] = len(course['extracted_text']) if course['extracted_text'] else 0

extraction_success = sum(1 for c in test_courses if c.get('extracted_text'))
print(f'Text extracted: {extraction_success}/{fetch_success}\n')

# Statistics
if extraction_success > 0:
    lengths = [c['text_length'] for c in test_courses if c.get('extracted_text')]
    print('Text length statistics:')
    print(f'  Min: {min(lengths)} characters')
    print(f'  Max: {max(lengths)} characters')
    print(f'  Mean: {int(sum(lengths)/len(lengths))} characters')
    print(f'  Median: {sorted(lengths)[len(lengths)//2]} characters')

# Summary
print('\n=== SUMMARY ===')
print(f'Total courses tested: {len(test_courses)}')
print(f'URL generation: {url_success}/{len(test_courses)} ({100*url_success//len(test_courses)}%)')
print(f'HTML fetch: {fetch_success}/{url_success} ({100*fetch_success//url_success if url_success > 0 else 0}%)')
print(f'Page verification: {verified}/{fetch_success} ({100*verified//fetch_success if fetch_success > 0 else 0}%)')
print(f'Text extraction: {extraction_success}/{fetch_success} ({100*extraction_success//fetch_success if fetch_success > 0 else 0}%)')

# Sample
if extraction_success > 0:
    sample = next(c for c in test_courses if c.get('extracted_text'))
    print('\n=== SAMPLE EXTRACTION ===')
    print(f'Course: {sample["code"]}')
    print(f'Text length: {sample["text_length"]} characters')
    print('First 500 characters:')
    print(sample['extracted_text'][:500])
    print('...\n')

# Save detailed results for later analysis
results_path = Path('data/nih_validation_results.txt')
with open(results_path, 'w') as f:
    f.write('NIH Validation Results\n')
    f.write('='*50 + '\n\n')
    for i, course in enumerate(test_courses):
        f.write(f'{i+1}. {course["code"]} ({course["year"]} Sem{course["sem"]})\n')
        f.write(f'   URL: {course["url"]}\n')
        f.write(f'   Fetch: {"OK" if course["html_success"] else "FAILED"}\n')
        if course.get('has_course_code'):
            f.write(f'   Course code found: YES\n')
        if course.get('extracted_text'):
            f.write(f'   Text extracted: {course["text_length"]} chars\n')
        if course.get('html_error'):
            f.write(f'   Error: {course["html_error"]}\n')
        f.write('\n')

print(f'\nDetailed results saved to {results_path}')
print('\nValidation complete!')
